In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

In [2]:
df = pd.read_csv('/Users/hitendraparanjpe/Downloads/dataset1-master/mushrooms.csv')
df

class cap-shape cap-surface cap-color bruises odor gill-attachment  \
0        p         x           s         n       t    p               f   
1        e         x           s         y       t    a               f   
2        e         b           s         w       t    l               f   
3        p         x           y         w       t    p               f   
4        e         x           s         g       f    n               f   
...    ...       ...         ...       ...     ...  ...             ...   
8119     e         k           s         n       f    n               a   
8120     e         x           s         n       f    n               a   
8121     e         f           s         n       f    n               a   
8122     p         k           y         n       f    y               f   
8123     e         x           s         n       f    n               a   

     gill-spacing gill-size gill-color  ... stalk-surface-below-ring  \
0               c         n          k  ...                        s   
1               c         b          k  ...                        s   
2               c         b          n  ...                        s   
3               c         n          n  ...                        s   
4               w         b          k  ...                        s   
...           ...       ...        ...  ...                      ...   
8119            c         b          y  ...                        s   
8120            c         b          y  ...                        s   
8121            c         b          n  ...                        s   
8122            c         n          b  ...                        k   
8123            c         b          y  ...                        s   

     stalk-color-above-ring stalk-color-below-ring veil-type veil-color  \
0                         w                      w         p          w   
1                         w                      w         p          w   
2                         w                      w         p          w   
3                         w                      w         p          w   
4                         w                      w         p          w   
...                     ...                    ...       ...        ...   
8119                      o                      o         p          o   
8120                      o                      o         p          n   
8121                      o                      o         p          o   
8122                      w                      w         p          w   
8123                      o                      o         p          o   

     ring-number ring-type spore-print-color population habitat  
0              o         p                 k          s       u  
1              o         p                 n          n       g  
2              o         p                 n          n       m  
3              o         p                 k          s       u  
4              o         e                 n          a       g  
...          ...       ...               ...        ...     ...  
8119           o         p                 b          c       l  
8120           o         p                 b          v       l  
8121           o         p                 b          c       l  
8122           o         e                 w          v       l  
8123           o         p                 o          c       l  

[8124 rows x 23 columns]

In [3]:
df.sample(20)

class cap-shape cap-surface cap-color bruises odor gill-attachment  \
5986     p         f           y         y       f    f               f   
2152     e         x           y         n       t    n               f   
213      e         x           f         g       f    n               f   
1999     e         x           s         n       f    n               f   
5        e         x           y         y       t    a               f   
7825     e         b           s         n       f    n               a   
7239     e         f           s         n       f    n               a   
5825     p         f           y         y       f    f               f   
4042     p         f           y         g       f    f               f   
3524     e         f           f         e       t    n               f   
8111     e         k           s         w       f    n               f   
1953     e         x           f         g       t    n               f   
7538     p         k           s         e       f    y               f   
7017     p         f           y         e       f    y               f   
5621     p         f           s         w       t    f               f   
3615     e         f           y         g       t    n               f   
208      e         b           y         w       t    l               f   
3875     e         x           f         g       t    n               f   
858      e         b           s         y       t    a               f   
2370     e         x           f         n       t    n               f   

     gill-spacing gill-size gill-color  ... stalk-surface-below-ring  \
5986            c         b          g  ...                        k   
2152            c         b          u  ...                        s   
213             c         n          n  ...                        s   
1999            w         b          k  ...                        f   
5               c         b          n  ...                        s   
7825            c         b          y  ...                        s   
7239            c         b          n  ...                        s   
5825            c         b          p  ...                        k   
4042            c         b          h  ...                        k   
3524            c         b          w  ...                        s   
8111            w         b          p  ...                        s   
1953            c         b          u  ...                        s   
7538            c         n          b  ...                        k   
7017            c         n          b  ...                        s   
5621            c         b          h  ...                        s   
3615            c         b          n  ...                        s   
208             c         b          n  ...                        s   
3875            c         b          w  ...                        s   
858             c         b          k  ...                        s   
2370            c         b          w  ...                        s   

     stalk-color-above-ring stalk-color-below-ring veil-type veil-color  \
5986                      b                      b         p          w   
2152                      p                      g         p          w   
213                       w                      w         p          w   
1999                      w                      w         p          w   
5                         w                      w         p          w   
7825                      o                      o         p          n   
7239                      o                      o         p          n   
5825                      n                      p         p          w   
4042                      b                      p         p          w   
3524                      w                      g         p          w   
8111                      w                      w         p          w   
1953            

# Inference

Each column of the dataset describes a feature of the mushroom.

All the columns in the dataset have categorical data.

"Class" column classifies the data into 2 catergories:

        1. e = Edible
        
        2. p = Poisonous
        

In [4]:
df['class'].value_counts()

e    4208
p    3916
Name: class, dtype: int64

In [5]:
df.describe()

class cap-shape cap-surface cap-color bruises  odor gill-attachment  \
count   8124      8124        8124      8124    8124  8124            8124   
unique     2         6           4        10       2     9               2   
top        e         x           y         n       f     n               f   
freq    4208      3656        3244      2284    4748  3528            7914   

       gill-spacing gill-size gill-color  ... stalk-surface-below-ring  \
count          8124      8124       8124  ...                     8124   
unique            2         2         12  ...                        4   
top               c         b          b  ...                        s   
freq           6812      5612       1728  ...                     4936   

       stalk-color-above-ring stalk-color-below-ring veil-type veil-color  \
count                    8124                   8124      8124       8124   
unique                      9                      9         1          4   
top                         w                      w         p          w   
freq                     4464                   4384      8124       7924   

       ring-number ring-type spore-print-color population habitat  
count         8124      8124              8124       8124    8124  
unique           3         5                 9          6       7  
top              o         p                 w          v       d  
freq          7488      3968              2388       4040    3148  

[4 rows x 23 columns]

This shows the number of unique values in each column of the dataset.

In the column veil-type, there is only One type of value 'p'. It occurs in all the 8124 rows. Hence, this column can be removed for developing our machine learning model.

In [6]:
df.isnull().sum() 

class                       0
cap-shape                   0
cap-surface                 0
cap-color                   0
bruises                     0
odor                        0
gill-attachment             0
gill-spacing                0
gill-size                   0
gill-color                  0
stalk-shape                 0
stalk-root                  0
stalk-surface-above-ring    0
stalk-surface-below-ring    0
stalk-color-above-ring      0
stalk-color-below-ring      0
veil-type                   0
veil-color                  0
ring-number                 0
ring-type                   0
spore-print-color           0
population                  0
habitat                     0
dtype: int64

There are no null values

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8124 entries, 0 to 8123
Data columns (total 23 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   class                     8124 non-null   object
 1   cap-shape                 8124 non-null   object
 2   cap-surface               8124 non-null   object
 3   cap-color                 8124 non-null   object
 4   bruises                   8124 non-null   object
 5   odor                      8124 non-null   object
 6   gill-attachment           8124 non-null   object
 7   gill-spacing              8124 non-null   object
 8   gill-size                 8124 non-null   object
 9   gill-color                8124 non-null   object
 10  stalk-shape               8124 non-null   object
 11  stalk-root                8124 non-null   object
 12  stalk-surface-above-ring  8124 non-null   object
 13  stalk-surface-below-ring  8124 non-null   object
 14  stalk-color-above-ring  

In [8]:
df = df.drop(columns=['veil-type'],axis=1)
df

class cap-shape cap-surface cap-color bruises odor gill-attachment  \
0        p         x           s         n       t    p               f   
1        e         x           s         y       t    a               f   
2        e         b           s         w       t    l               f   
3        p         x           y         w       t    p               f   
4        e         x           s         g       f    n               f   
...    ...       ...         ...       ...     ...  ...             ...   
8119     e         k           s         n       f    n               a   
8120     e         x           s         n       f    n               a   
8121     e         f           s         n       f    n               a   
8122     p         k           y         n       f    y               f   
8123     e         x           s         n       f    n               a   

     gill-spacing gill-size gill-color  ... stalk-surface-above-ring  \
0               c         n          k  ...                        s   
1               c         b          k  ...                        s   
2               c         b          n  ...                        s   
3               c         n          n  ...                        s   
4               w         b          k  ...                        s   
...           ...       ...        ...  ...                      ...   
8119            c         b          y  ...                        s   
8120            c         b          y  ...                        s   
8121            c         b          n  ...                        s   
8122            c         n          b  ...                        s   
8123            c         b          y  ...                        s   

     stalk-surface-below-ring stalk-color-above-ring stalk-color-below-ring  \
0                           s                      w                      w   
1                           s                      w                      w   
2                           s                      w                      w   
3                           s                      w                      w   
4                           s                      w                      w   
...                       ...                    ...                    ...   
8119                        s                      o                      o   
8120                        s                      o                      o   
8121                        s                      o                      o   
8122                        k                      w                      w   
8123                        s                      o                      o   

     veil-color ring-number ring-type spore-print-color population habitat  
0             w           o         p                 k          s       u  
1             w           o         p                 n          n       g  
2             w           o         p                 n          n       m  
3             w           o         p                 k          s       u  
4             w           o         e                 n          a       g  
...         ...         ...       ...               ...        ...     ...  
8119          o           o         p                 b          c       l  
8120          n           o         p                 b          v       l  
8121          o           o         p                 b          c       l  
8122          w           o         e                 w          v       l  
8123          o           o         p                 o          c       l  

[8124 rows x 22 columns]

# Inference on Data Type
We can see the data is not integer/float type. Since the Data is Object type, we will need to convert it to Category type to analyse further.

In [9]:
from sklearn.preprocessing import LabelEncoder

In [10]:
le = LabelEncoder()

In [11]:
for i in df.columns:
    df[i] = le.fit_transform(df[i])

In [12]:
df

class  cap-shape  cap-surface  cap-color  bruises  odor  \
0         1          5            2          4        1     6   
1         0          5            2          9        1     0   
2         0          0            2          8        1     3   
3         1          5            3          8        1     6   
4         0          5            2          3        0     5   
...     ...        ...          ...        ...      ...   ...   
8119      0          3            2          4        0     5   
8120      0          5            2          4        0     5   
8121      0          2            2          4        0     5   
8122      1          3            3          4        0     8   
8123      0          5            2          4        0     5   

      gill-attachment  gill-spacing  gill-size  gill-color  ...  \
0                   1             0          1           4  ...   
1                   1             0          0           4  ...   
2                   1             0          0           5  ...   
3                   1             0          1           5  ...   
4                   1             1          0           4  ...   
...               ...           ...        ...         ...  ...   
8119                0             0          0          11  ...   
8120                0             0          0          11  ...   
8121                0             0          0           5  ...   
8122                1             0          1           0  ...   
8123                0             0          0          11  ...   

      stalk-surface-above-ring  stalk-surface-below-ring  \
0                            2                         2   
1                            2                         2   
2                            2                         2   
3                            2                         2   
4                            2                         2   
...                        ...                       ...   
8119                         2                         2   
8120                         2                         2   
8121                         2                         2   
8122                         2                         1   
8123                         2                         2   

      stalk-color-above-ring  stalk-color-below-ring  veil-color  ring-number  \
0                          7                       7           2            1   
1                          7                       7           2            1   
2                          7                       7           2            1   
3                          7                       7           2            1   
4                          7                       7           2            1   
...                      ...                     ...         ...          ...   
8119                       5                       5           1            1   
8120                       5                       5           0            1   
8121                       5                       5           1            1   
8122                       7                       7           2            1   
8123                       5                       5           1            1   

      ring-type  spore-print-color  population  habitat  
0             4                  2           3        5  
1             4                  3           2        1  
2             4                  3           2        3  
3             4                  2           3        5  
4             0                  3           0        1  
...         ...                ...         ...      ...  
8119          4                  0           1        2  
8120          4                  0           4        2  
8121          4                  0           1        2  
8122          0                  7           4        2  
8123          4                  4           1        2  

[8124 rows x 22 columns]

In [13]:
x = df.drop(columns=['class'],axis=1)
y = df['class']

In [14]:
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report

In [15]:
x_train, x_test, y_train, y_test = train_test_split(x,y, test_size = 0.25, random_state = 355)

In [16]:
dtree = DecisionTreeClassifier()

In [17]:
dtree.fit(x_train,y_train)

DecisionTreeClassifier()

In [18]:
y_pred = dtree.predict(x_test)

In [19]:
cm = confusion_matrix(y_test, y_pred)
cm

array([[1058,    0],
       [   0,  973]])

In [20]:
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00      1058
           1       1.00      1.00      1.00       973

    accuracy                           1.00      2031
   macro avg       1.00      1.00      1.00      2031
weighted avg       1.00      1.00      1.00      2031



The Decision Tree Model has given 100% accuracy.